File: clients/go-sdk/client/principals.go
package client

import (
	"context"
	"fmt"
)

// ─── Request DTOs ────────────────────────────────────────────────────

// CreateUserRequest — POST /api/principals/users.
type CreateUserRequest struct {
	Email    string `json:"email"`
	Password string `json:"password,omitempty"`
	Name     string `json:"name"`
	ClientID string `json:"clientId,omitempty"`
	// EnforcePasswordComplexity: pass *false to skip the platform's
	// upper/lower/digit/special rules (2-char minimum still applies).
	// Use when your application enforces its own password policy.
	EnforcePasswordComplexity *bool `json:"enforcePasswordComplexity,omitempty"`
}

// UpdatePrincipalRequest — PUT /api/principals/{id}.
type UpdatePrincipalRequest struct {
	Name      *string `json:"name,omitempty"`
	FirstName *string `json:"firstName,omitempty"`
	LastName  *string `json:"lastName,omitempty"`
	Active    *bool   `json:"active,omitempty"`
}

// ResetPasswordRequest — POST /api/principals/{id}/reset-password.
type ResetPasswordRequest struct {
	NewPassword               string `json:"newPassword"`
	EnforcePasswordComplexity *bool  `json:"enforcePasswordComplexity,omitempty"`
}

// AssignRoleRequest — body for POST /api/principals/{id}/roles. The
// backend expects { role }, not { roleName }.
type AssignRoleRequest struct {
	Role string `json:"role"`
}

// ReplaceRolesRequest — body for PUT /api/principals/{id}/roles.
type ReplaceRolesRequest struct {
	Roles []string `json:"roles"`
}

// GrantClientAccessRequest — body for POST /api/principals/{id}/client-access.
type GrantClientAccessRequest struct {
	ClientID string `json:"clientId"`
}

// PrincipalFilters — query parameters for GET /api/principals.
type PrincipalFilters struct {
	ClientID string
	Type     string
	Active   string
	Email    string
}

// ─── Response DTOs ───────────────────────────────────────────────────

// PrincipalResponse is the platform's principal aggregate.
type PrincipalResponse struct {
	ID                string   `json:"id"`
	PrincipalType     string   `json:"type"`
	Scope             string   `json:"scope"`
	ClientID          string   `json:"clientId,omitempty"`
	Name              string   `json:"name"`
	Active            bool     `json:"active"`
	Email             string   `json:"email,omitempty"`
	IdpType           string   `json:"idpType,omitempty"`
	Roles             []string `json:"roles,omitempty"`
	IsAnchorUser      bool     `json:"isAnchorUser,omitempty"`
	GrantedClientIDs  []string `json:"grantedClientIds,omitempty"`
	CreatedAt         string   `json:"createdAt"`
	UpdatedAt         string   `json:"updatedAt"`
}

// PrincipalListResponse — GET /api/principals.
type PrincipalListResponse struct {
	Principals []PrincipalResponse `json:"principals"`
	Total      uint64              `json:"total,omitempty"`
}

// PrincipalRoleResponse — one item from GET /api/principals/{id}/roles.
// Matches the platform's RoleAssignmentDto.
type PrincipalRoleResponse struct {
	ID               string `json:"id"`
	RoleName         string `json:"roleName"`
	AssignmentSource string `json:"assignmentSource"`
	AssignedAt       string `json:"assignedAt"`
}

// PrincipalRoleListResponse — GET /api/principals/{id}/roles.
type PrincipalRoleListResponse struct {
	Roles []PrincipalRoleResponse `json:"roles"`
}

// ClientAccessGrantResponse — one item from GET /api/principals/{id}/client-access.
type ClientAccessGrantResponse struct {
	ID        string `json:"id"`
	ClientID  string `json:"clientId"`
	GrantedAt string `json:"grantedAt"`
	ExpiresAt string `json:"expiresAt,omitempty"`
}

// ClientAccessListResponse — GET /api/principals/{id}/client-access.
type ClientAccessListResponse struct {
	Grants []ClientAccessGrantResponse `json:"grants"`
}

// ─── Sync ────────────────────────────────────────────────────────────

// SyncPrincipalItem — one entry in the per-app sync payload. Matched by
// email. Role names omit the `<app>:` prefix — the platform adds it.
type SyncPrincipalItem struct {
	Email  string   `json:"email"`
	Name   string   `json:"name"`
	Roles  []string `json:"roles,omitempty"`
	Active bool     `json:"active"`
}

// SyncPrincipalsRequest — body for the per-app sync endpoint.
type SyncPrincipalsRequest struct {
	Principals []SyncPrincipalItem `json:"principals"`
}

// ─── Resource ────────────────────────────────────────────────────────

// PrincipalsResource — /api/principals/*.
type PrincipalsResource struct {
	c *FlowCatalystClient
}

// CreateUser — POST /api/principals/users.
func (r *PrincipalsResource) CreateUser(ctx context.Context, req *CreateUserRequest) (*PrincipalResponse, error) {
	var out PrincipalResponse
	if err := r.c.Post(ctx, "/api/principals/users", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List — GET /api/principals with optional filters.
func (r *PrincipalsResource) List(ctx context.Context, filters *PrincipalFilters) (*PrincipalListResponse, error) {
	q := ""
	if filters != nil {
		q = NewQuery().
			String("clientId", filters.ClientID).
			String("type", filters.Type).
			String("active", filters.Active).
			String("email", filters.Email).
			Encode()
	}
	var out PrincipalListResponse
	if err := r.c.Get(ctx, "/api/principals"+q, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get — GET /api/principals/{id}.
func (r *PrincipalsResource) Get(ctx context.Context, id string) (*PrincipalResponse, error) {
	var out PrincipalResponse
	if err := r.c.Get(ctx, "/api/principals/"+id, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// FindByEmail — convenience over List with the email filter. Returns
// every principal the caller is authorised to see whose email matches
// exactly (case-insensitive). Callers should pick the expected one by
// email rather than assuming index 0.
func (r *PrincipalsResource) FindByEmail(ctx context.Context, email string) (*PrincipalListResponse, error) {
	return r.List(ctx, &PrincipalFilters{Email: email})
}

// Update — PUT /api/principals/{id}.
func (r *PrincipalsResource) Update(ctx context.Context, id string, req *UpdatePrincipalRequest) (*PrincipalResponse, error) {
	var out PrincipalResponse
	if err := r.c.Put(ctx, "/api/principals/"+id, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Activate — POST /api/principals/{id}/activate. Platform returns
// { message } only; call Get(id) for the refreshed record.
func (r *PrincipalsResource) Activate(ctx context.Context, id string) error {
	return r.c.Post(ctx, "/api/principals/"+id+"/activate", nil, nil)
}

// Deactivate — POST /api/principals/{id}/deactivate.
func (r *PrincipalsResource) Deactivate(ctx context.Context, id string) error {
	return r.c.Post(ctx, "/api/principals/"+id+"/deactivate", nil, nil)
}

// Roles — GET /api/principals/{id}/roles.
func (r *PrincipalsResource) Roles(ctx context.Context, id string) (*PrincipalRoleListResponse, error) {
	var out PrincipalRoleListResponse
	if err := r.c.Get(ctx, "/api/principals/"+id+"/roles", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// AddRole — POST /api/principals/{id}/roles (additive — keeps existing
// roles). Distinct from SetRoles which replaces the full set.
func (r *PrincipalsResource) AddRole(ctx context.Context, id, roleName string) error {
	return r.c.Post(ctx, "/api/principals/"+id+"/roles", &AssignRoleRequest{Role: roleName}, nil)
}

// RemoveRole — DELETE /api/principals/{id}/roles/{roleName}.
func (r *PrincipalsResource) RemoveRole(ctx context.Context, id, roleName string) error {
	return r.c.Delete(ctx, fmt.Sprintf("/api/principals/%s/roles/%s", id, roleName), nil)
}

// SetRoles — PUT /api/principals/{id}/roles. Replaces the full set.
func (r *PrincipalsResource) SetRoles(ctx context.Context, id string, roles []string) error {
	return r.c.Put(ctx, "/api/principals/"+id+"/roles", &ReplaceRolesRequest{Roles: roles}, nil)
}

// ClientAccessGrants — GET /api/principals/{id}/client-access.
func (r *PrincipalsResource) ClientAccessGrants(ctx context.Context, id string) (*ClientAccessListResponse, error) {
	var out ClientAccessListResponse
	if err := r.c.Get(ctx, "/api/principals/"+id+"/client-access", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GrantClientAccess — POST /api/principals/{id}/client-access.
func (r *PrincipalsResource) GrantClientAccess(ctx context.Context, principalID, clientID string) error {
	return r.c.Post(ctx, "/api/principals/"+principalID+"/client-access",
		&GrantClientAccessRequest{ClientID: clientID}, nil)
}

// RevokeClientAccess — DELETE /api/principals/{id}/client-access/{clientId}.
func (r *PrincipalsResource) RevokeClientAccess(ctx context.Context, principalID, clientID string) error {
	return r.c.Delete(ctx,
		fmt.Sprintf("/api/principals/%s/client-access/%s", principalID, clientID), nil)
}

// ResetPassword — POST /api/principals/{id}/reset-password.
func (r *PrincipalsResource) ResetPassword(ctx context.Context, principalID string, req *ResetPasswordRequest) error {
	return r.c.Post(ctx, "/api/principals/"+principalID+"/reset-password", req, nil)
}

// Sync — POST /api/applications/{appCode}/principals/sync. When
// removeUnlisted is true the platform strips SDK-sourced role
// assignments from principals not in the list (principals themselves
// are never deleted by sync).
func (r *PrincipalsResource) Sync(ctx context.Context, appCode string, req *SyncPrincipalsRequest, removeUnlisted bool) (*SyncResult, error) {
	q := ""
	if removeUnlisted {
		q = "?removeUnlisted=true"
	}
	var out SyncResult
	if err := r.c.Post(ctx, fmt.Sprintf("/api/applications/%s/principals/sync%s", appCode, q), req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: crates/fc-sdk/src/client/principals.rs
//! Principal (user/service) management operations.

use super::{ClientError, FlowCatalystClient};
use serde::{Deserialize, Serialize};

/// Request to create a user principal.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CreateUserRequest {
    pub email: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password: Option<String>,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_id: Option<String>,
    /// When set to `false`, the platform skips its password complexity rules
    /// (uppercase/lowercase/digit/special) and only enforces a 2-character minimum.
    /// Use when your application enforces its own password policy.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enforce_password_complexity: Option<bool>,
}

/// Request to reset a principal's password via the admin API.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ResetPasswordRequest {
    pub new_password: String,
    /// When set to `false`, the platform skips its password complexity rules
    /// (uppercase/lowercase/digit/special) and only enforces a 2-character minimum.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enforce_password_complexity: Option<bool>,
}

/// Request to update a principal.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdatePrincipalRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub first_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active: Option<bool>,
}

/// Filters for listing principals.
#[derive(Debug, Clone, Default)]
pub struct PrincipalFilters {
    pub client_id: Option<String>,
    pub r#type: Option<String>,
    pub active: Option<String>,
    pub email: Option<String>,
}

/// Principal response from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrincipalResponse {
    pub id: String,
    #[serde(rename = "type")]
    pub principal_type: String,
    pub scope: String,
    #[serde(default)]
    pub client_id: Option<String>,
    pub name: String,
    pub active: bool,
    #[serde(default)]
    pub email: Option<String>,
    #[serde(default)]
    pub idp_type: Option<String>,
    #[serde(default)]
    pub roles: Vec<String>,
    #[serde(default)]
    pub is_anchor_user: bool,
    #[serde(default)]
    pub granted_client_ids: Vec<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Role assignment returned by GET /api/principals/{id}/roles.
///
/// Matches the platform's `RoleAssignmentDto` — uses `roleName` and
/// `assignmentSource`, not `name` and `source`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrincipalRoleResponse {
    pub id: String,
    pub role_name: String,
    pub assignment_source: String,
    pub assigned_at: String,
}

/// Client access grant for a principal.
///
/// Matches the platform's `ClientAccessGrantResponse` — `id`, `clientId`,
/// `grantedAt`, optional `expiresAt`. No client name / identifier fields
/// (those were in an older SDK version that didn't reflect the backend).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientAccessGrantResponse {
    pub id: String,
    pub client_id: String,
    pub granted_at: String,
    #[serde(default)]
    pub expires_at: Option<String>,
}

/// Request to assign a single role (additive — keeps existing roles).
///
/// The backend expects `{ "role": "..." }`, not `{ "roleName": "..." }`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AssignRoleRequest {
    pub role: String,
}

/// Request to replace all roles.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReplaceRolesRequest {
    pub roles: Vec<String>,
}

/// Request to grant client access.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GrantClientAccessRequest {
    pub client_id: String,
}

/// Paginated list of principals — `GET /api/principals`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrincipalListResponse {
    pub principals: Vec<PrincipalResponse>,
    #[serde(default)]
    pub total: u64,
}

/// List of role assignments for a principal — `GET /api/principals/{id}/roles`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrincipalRoleListResponse {
    pub roles: Vec<PrincipalRoleResponse>,
}

/// List of client access grants — `GET /api/principals/{id}/client-access`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientAccessListResponse {
    pub grants: Vec<ClientAccessGrantResponse>,
}

/// Request body for the per-resource sync endpoint.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncPrincipalsRequest {
    pub principals: Vec<SyncPrincipalItem>,
}

/// A principal item for sync. Matched by email.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncPrincipalItem {
    pub email: String,
    pub name: String,
    /// Role short names (without the `<app>:` prefix — the platform adds it).
    #[serde(default)]
    pub roles: Vec<String>,
    /// Defaults to `true` server-side when omitted.
    #[serde(default = "default_true")]
    pub active: bool,
}

fn default_true() -> bool {
    true
}

/// Principals resource accessor — created via [`FlowCatalystClient::principals`].
pub struct Principals<'a> {
    pub(crate) client: &'a FlowCatalystClient,
}

impl Principals<'_> {
    /// Create a new user principal.
    pub async fn create_user(
        &self,
        req: &CreateUserRequest,
    ) -> Result<PrincipalResponse, ClientError> {
        self.client.post("/api/principals/users", req).await
    }

    /// List principals with optional filters.
    pub async fn list(
        &self,
        filters: &PrincipalFilters,
    ) -> Result<PrincipalListResponse, ClientError> {
        let mut params = Vec::new();
        if let Some(ref cid) = filters.client_id {
            params.push(format!("clientId={}", cid));
        }
        if let Some(ref t) = filters.r#type {
            params.push(format!("type={}", t));
        }
        if let Some(ref a) = filters.active {
            params.push(format!("active={}", a));
        }
        if let Some(ref e) = filters.email {
            params.push(format!("email={}", e));
        }
        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };
        self.client.get(&format!("/api/principals{}", query)).await
    }

    /// Get a principal by ID.
    pub async fn get(&self, id: &str) -> Result<PrincipalResponse, ClientError> {
        self.client.get(&format!("/api/principals/{}", id)).await
    }

    /// Find principals by email.
    ///
    /// The result still contains every principal the caller is authorised to
    /// see whose email matches exactly (case-insensitive) — callers should
    /// pick the expected one by `email` rather than assuming index 0.
    pub async fn find_by_email(
        &self,
        email: &str,
    ) -> Result<PrincipalListResponse, ClientError> {
        self.list(&PrincipalFilters {
            email: Some(email.to_string()),
            ..Default::default()
        })
        .await
    }

    /// Update a principal.
    pub async fn update(
        &self,
        id: &str,
        req: &UpdatePrincipalRequest,
    ) -> Result<PrincipalResponse, ClientError> {
        self.client
            .put(&format!("/api/principals/{}", id), req)
            .await
    }

    /// Activate a principal.
    ///
    /// The platform returns `{ "message": "..." }` only. Call `get(id)` if
    /// you need the refreshed record.
    pub async fn activate(&self, id: &str) -> Result<(), ClientError> {
        let _: serde_json::Value = self
            .client
            .post_action(&format!("/api/principals/{}/activate", id))
            .await?;
        Ok(())
    }

    /// Deactivate a principal. The platform returns a message only.
    pub async fn deactivate(&self, id: &str) -> Result<(), ClientError> {
        let _: serde_json::Value = self
            .client
            .post_action(&format!("/api/principals/{}/deactivate", id))
            .await?;
        Ok(())
    }

    /// Get roles assigned to a principal.
    pub async fn roles(
        &self,
        id: &str,
    ) -> Result<PrincipalRoleListResponse, ClientError> {
        self.client
            .get(&format!("/api/principals/{}/roles", id))
            .await
    }

    /// Add a single role to a principal (additive — keeps existing roles).
    ///
    /// Distinct from [`set_roles`] which replaces the full set. Renamed from
    /// the original `assign_role` / `assign_roles` pair to make the
    /// additive-vs-replace distinction visible at the call site.
    pub async fn add_role(&self, id: &str, role_name: &str) -> Result<(), ClientError> {
        let body = AssignRoleRequest {
            role: role_name.to_string(),
        };
        let _: serde_json::Value = self
            .client
            .post(&format!("/api/principals/{}/roles", id), &body)
            .await?;
        Ok(())
    }

    /// Remove a role from a principal.
    pub async fn remove_role(&self, id: &str, role_name: &str) -> Result<(), ClientError> {
        self.client
            .delete_req(&format!("/api/principals/{}/roles/{}", id, role_name))
            .await
    }

    /// Replace all roles on a principal with the given set.
    pub async fn set_roles(&self, id: &str, roles: Vec<String>) -> Result<(), ClientError> {
        let body = ReplaceRolesRequest { roles };
        let _: serde_json::Value = self
            .client
            .put(&format!("/api/principals/{}/roles", id), &body)
            .await?;
        Ok(())
    }

    /// Get client access grants for a principal.
    pub async fn client_access_grants(
        &self,
        id: &str,
    ) -> Result<ClientAccessListResponse, ClientError> {
        self.client
            .get(&format!("/api/principals/{}/client-access", id))
            .await
    }

    /// Grant client access to a principal.
    pub async fn grant_client_access(
        &self,
        principal_id: &str,
        client_id: &str,
    ) -> Result<(), ClientError> {
        let body = GrantClientAccessRequest {
            client_id: client_id.to_string(),
        };
        let _: serde_json::Value = self
            .client
            .post(
                &format!("/api/principals/{}/client-access", principal_id),
                &body,
            )
            .await?;
        Ok(())
    }

    /// Revoke client access from a principal.
    pub async fn revoke_client_access(
        &self,
        principal_id: &str,
        client_id: &str,
    ) -> Result<(), ClientError> {
        self.client
            .delete_req(&format!(
                "/api/principals/{}/client-access/{}",
                principal_id, client_id
            ))
            .await
    }

    /// Reset a principal's password via the admin API.
    pub async fn reset_password(
        &self,
        principal_id: &str,
        req: &ResetPasswordRequest,
    ) -> Result<(), ClientError> {
        let _: serde_json::Value = self
            .client
            .post(
                &format!("/api/principals/{}/reset-password", principal_id),
                req,
            )
            .await?;
        Ok(())
    }

    /// Sync principals for an application — declarative reconciliation
    /// against `POST /api/applications/{appCode}/principals/sync`.
    ///
    /// When `remove_unlisted` is true the platform strips SDK-sourced role
    /// assignments from principals not in the list (principals themselves
    /// are never deleted by sync).
    pub async fn sync(
        &self,
        app_code: &str,
        req: &SyncPrincipalsRequest,
        remove_unlisted: bool,
    ) -> Result<crate::client::SyncResult, ClientError> {
        let query = if remove_unlisted {
            "?removeUnlisted=true"
        } else {
            ""
        };
        self.client
            .post(
                &format!("/api/applications/{}/principals/sync{}", app_code, query),
                req,
            )
            .await
    }
}
